<h1><center>Analyzing Customer Orders Using Python</center></h1>

<h2>1. Store customer orders</h2>

<h4>Create a list of customer names</h4>

In [41]:
#Create a list of customer names
customers = ["Adhir Samal","Ashish Pandey","Vipin Kumar","Ranjeet Kumar","Sajal Singhal",
             "Akhilesh Kumar","Suvankar Chakraborty","Suvankar Adhikary","Rahul Kumar Prajapati",
             "Sagar Jeur","Vipin Kumar","Ashish Kumar","Haridas Balraj","Kundan Kumar",
             "Atif Siddiqui","Firdoush Jabee","Jay Prakash Kumar","Ramkrishna Das Adhikary","Ranjeet Mandal"]

<h4>order details (customer name, product, price, category)</h4>

In [42]:
#Store each customer's order details (customer name, product, price, category) as tuples inside a list

orders = [
    ("Adhir Samal", "Laptops", 90, "Electronics"),
    ("Ashish Pandey", "Tablets", 70, "Electronics"),
    ("Vipin Kumar", "Hoodies", 70, "Clothing"),
    ("Ranjeet Kumar", "Curtains", 65, "Home Essentials"),
    ("Sajal Singhal", "Cameras", 65, "Electronics"),
    ("Akhilesh Kumar", "Jeans", 85, "Clothing"),
    ("Suvankar Chakraborty", "Home organizers", 40, "Home Essentials"),
    ("Suvankar Adhikary", "Laptops", 90, "Electronics"),
    ("Rahul Kumar Prajapati", "Shirts", 65, "Clothing"),
    ("Sagar Jeur", "Bedsheets", 85, "Home Essentials"),
    ("Vipin Kumar", "Tablets", 70, "Electronics"),
    ("Ashish Kumar", "Hoodies", 70, "Clothing"),
    ("Akhilesh Kumar", "Curtains", 65, "Home Essentials"),
    ("Haridas Balraj", "Home organizers", 40, "Home Essentials"),
    ("Kundan Kumar", "Jeans", 85, "Clothing"),
    ("Atif Siddiqui", "Bedsheets", 85, "Home Essentials"),
    ("Adhir Samal", "Curtains", 65, "Home Essentials"),
    ("Firdoush Jabee", "Shirts", 65, "Clothing"),
    ("Jay Prakash Kumar", "Jeans", 85, "Clothing"),
    ("Ashish Pandey", "Home organizers", 40, "Home Essentials"),
    ("Akhilesh Kumar", "Cameras", 65, "Electronics"),
    ("Suvankar Chakraborty", "Bedsheets", 85, "Home Essentials"),
    ("Ramkrishna Das Adhikary", "Tablets", 70, "Electronics"),
    ("Ranjeet Mandal", "Cameras", 65, "Electronics")
]

<h4>Dictionary: keys are customer names and values are lists of ordered products</h4>

In [43]:
# Use a dictionary where keys are customer names and values are lists of ordered products

customer_orders={}

for cust_name, product, price, category in orders:
    
    if cust_name not in customer_orders:
        customer_orders[cust_name] = [product]
    else:
        customer_orders[cust_name].append(product)

print(customer_orders)

{'Adhir Samal': ['Laptops', 'Curtains'], 'Ashish Pandey': ['Tablets', 'Home organizers'], 'Vipin Kumar': ['Hoodies', 'Tablets'], 'Ranjeet Kumar': ['Curtains'], 'Sajal Singhal': ['Cameras'], 'Akhilesh Kumar': ['Jeans', 'Curtains', 'Cameras'], 'Suvankar Chakraborty': ['Home organizers', 'Bedsheets'], 'Suvankar Adhikary': ['Laptops'], 'Rahul Kumar Prajapati': ['Shirts'], 'Sagar Jeur': ['Bedsheets'], 'Ashish Kumar': ['Hoodies'], 'Haridas Balraj': ['Home organizers'], 'Kundan Kumar': ['Jeans'], 'Atif Siddiqui': ['Bedsheets'], 'Firdoush Jabee': ['Shirts'], 'Jay Prakash Kumar': ['Jeans'], 'Ramkrishna Das Adhikary': ['Tablets'], 'Ranjeet Mandal': ['Cameras']}


<h2>2. Classify products by category</h2>

<h4>Map each product to its respective category</h4>

In [44]:
# Use a dictionary to map each product to its respective category

product_category={}

for cust_name, product, price, category in orders:
    
    if product not in product_category:
        product_category[product] = category

print(product_category)

{'Laptops': 'Electronics', 'Tablets': 'Electronics', 'Hoodies': 'Clothing', 'Curtains': 'Home Essentials', 'Cameras': 'Electronics', 'Jeans': 'Clothing', 'Home organizers': 'Home Essentials', 'Shirts': 'Clothing', 'Bedsheets': 'Home Essentials'}


<h4>Unique product categories</h4>

In [45]:
# Create a set of unique product categories

categories=set()

for cust_name, product, price, category in orders:
        categories.add(category)

print(categories)

{'Home Essentials', 'Clothing', 'Electronics'}


<h4>Display all available product categories</h4>

In [46]:
# Display all available product categories

print("All available product categories:")

for cat in categories:
    print(cat)

All available product categories:
Home Essentials
Clothing
Electronics


<h4>Display products under each category</h4>

In [47]:
category_products={}

for cust_name, product, price, category in orders:
    
    if category not in category_products:
        category_products[category]= [product]
    else:
        if product not in category_products[category]:
            category_products[category].append(product)

print(category_products)

{'Electronics': ['Laptops', 'Tablets', 'Cameras'], 'Clothing': ['Hoodies', 'Jeans', 'Shirts'], 'Home Essentials': ['Curtains', 'Home organizers', 'Bedsheets']}


<h2> 3. Analyze customer orders </h2>

<h4>Total amount each customer spends</h4>

In [48]:
# Use a loop to calculate the total amount each customer spends

customer_totalsales={}

for cust_name, product, price, category in orders:
    
    if cust_name not in customer_totalsales:
        customer_totalsales[cust_name]= price
    else:
        customer_totalsales[cust_name]+=price

print(customer_totalsales)


{'Adhir Samal': 155, 'Ashish Pandey': 110, 'Vipin Kumar': 140, 'Ranjeet Kumar': 65, 'Sajal Singhal': 65, 'Akhilesh Kumar': 215, 'Suvankar Chakraborty': 125, 'Suvankar Adhikary': 90, 'Rahul Kumar Prajapati': 65, 'Sagar Jeur': 85, 'Ashish Kumar': 70, 'Haridas Balraj': 40, 'Kundan Kumar': 85, 'Atif Siddiqui': 85, 'Firdoush Jabee': 65, 'Jay Prakash Kumar': 85, 'Ramkrishna Das Adhikary': 70, 'Ranjeet Mandal': 65}


<h4>customer classification based on the total sale</h4>
<ul><li>If the total purchase value is above 100, classify the customer as a high-value buyer</li>
<li>If it is between 50 and 100, classify the customer as a moderate buyer</li>
<li>If it is below 50, classify them as a low-value buyer</li></ul>

In [49]:
buyer_classification={}

for customer_name, total_sale in customer_totalsales.items():
    
    if total_sale > 100:
        category = "High-Value Buyer"
    elif 50 <= total_sale <= 100:
        category = "Moderate Buyer"
    else:
        category = "Low-Value Buyer" 
    buyer_classification[customer_name]= category
    
print(buyer_classification)

{'Adhir Samal': 'High-Value Buyer', 'Ashish Pandey': 'High-Value Buyer', 'Vipin Kumar': 'High-Value Buyer', 'Ranjeet Kumar': 'Moderate Buyer', 'Sajal Singhal': 'Moderate Buyer', 'Akhilesh Kumar': 'High-Value Buyer', 'Suvankar Chakraborty': 'High-Value Buyer', 'Suvankar Adhikary': 'Moderate Buyer', 'Rahul Kumar Prajapati': 'Moderate Buyer', 'Sagar Jeur': 'Moderate Buyer', 'Ashish Kumar': 'Moderate Buyer', 'Haridas Balraj': 'Low-Value Buyer', 'Kundan Kumar': 'Moderate Buyer', 'Atif Siddiqui': 'Moderate Buyer', 'Firdoush Jabee': 'Moderate Buyer', 'Jay Prakash Kumar': 'Moderate Buyer', 'Ramkrishna Das Adhikary': 'Moderate Buyer', 'Ranjeet Mandal': 'Moderate Buyer'}


<h2>4. Generate business insights</h2>

<h4>Total revenue per product category</h4>

In [10]:
#  Calculate the total revenue per product category and store it in a dictionary

category_totalsales={}

for cust_name, product, price, category in orders:
    
    if category not in category_totalsales:
        category_totalsales[category]= price
    else:
        category_totalsales[category]+=price

print(category_totalsales)


{'Electronics': 585, 'Clothing': 525, 'Home Essentials': 570}


<h4>List of Unique Products</h4>

In [11]:
#  Extract unique products from all orders using a set

products=set()

for cust_name, product, price, category in orders:
        products.add(product)

print(products)


{'Curtains', 'Shirts', 'Hoodies', 'Bedsheets', 'Tablets', 'Jeans', 'Home organizers', 'Cameras', 'Laptops'}


<h4>All customers who purchased Fruits</h4>

In [14]:
# Use a list comprehension to find all customers who purchased Electronics

customers_electronics=[]

for cust_name, product, price, category in orders:
    
    if category is "Electronics":
        customers_electronics.append(cust_name)

print(customers_electronics)
        

['Adhir Samal', 'Ashish Pandey', 'Sajal Singhal', 'Suvankar Adhikary', 'Vipin Kumar', 'Akhilesh Kumar', 'Ramkrishna Das Adhikary', 'Ranjeet Mandal']


<h4>Top 3 highest-spending customers</h4>

In [15]:
#Identify the top three highest-spending customers using sorting

top3_customers=dict(list(customer_totalsales.items())[:3])

for customer in top3_customers:
    print(f"{customer} → Total: {customer_totalsales[customer]}")

Adhir Samal → Total: 155
Ashish Pandey → Total: 110
Vipin Kumar → Total: 140


<h2>5. Organize and display data</h2>

<h4>Each customer’s total spending and their classification</h4>

In [16]:
# Print a summary of each customer’s total spending and their classification
customer_totalsales = dict(sorted(customer_totalsales.items(), key=lambda item: item[1], reverse=True))

for customer in customer_totalsales:
    print(f"{customer} → Total: {customer_totalsales[customer]} → {buyer_classification[customer]}")

Akhilesh Kumar → Total: 215 → High-Value Buyer
Adhir Samal → Total: 155 → High-Value Buyer
Vipin Kumar → Total: 140 → High-Value Buyer
Suvankar Chakraborty → Total: 125 → High-Value Buyer
Ashish Pandey → Total: 110 → High-Value Buyer
Suvankar Adhikary → Total: 90 → Moderate Buyer
Sagar Jeur → Total: 85 → Moderate Buyer
Kundan Kumar → Total: 85 → Moderate Buyer
Atif Siddiqui → Total: 85 → Moderate Buyer
Jay Prakash Kumar → Total: 85 → Moderate Buyer
Ashish Kumar → Total: 70 → Moderate Buyer
Ramkrishna Das Adhikary → Total: 70 → Moderate Buyer
Ranjeet Kumar → Total: 65 → Moderate Buyer
Sajal Singhal → Total: 65 → Moderate Buyer
Rahul Kumar Prajapati → Total: 65 → Moderate Buyer
Firdoush Jabee → Total: 65 → Moderate Buyer
Ranjeet Mandal → Total: 65 → Moderate Buyer
Haridas Balraj → Total: 40 → Low-Value Buyer


<h4>Customers who purchased from multiple categories</h4>

In [17]:
# Use set operations to find customers who purchased from multiple categories
customers_mul_categories={}
categroies={}
for cust_name, product, price, category in orders:
        if cust_name not in categroies:
            categroies[cust_name]= [category]
        else:
            if category not in categroies[cust_name]:
                categroies[cust_name].append(category)
    
        if(len(categroies[cust_name])>=2):
            customers_mul_categories[cust_name]=categroies[cust_name]

if(len(customers_mul_categories)>=1):
    print("\nCustomers who purchased from multiple categories:")
    for cust_name in customers_mul_categories:
        print(cust_name)


Customers who purchased from multiple categories:
Vipin Kumar
Akhilesh Kumar
Adhir Samal
Ashish Pandey


<h4>Common customers who bought both Electronics and Clothing</h4>

In [19]:
# Identify common customers who bought both Beverages and Grains

customers_electronics_categories=set()
customers_clothing_categories=set()

for cust_name, product, price, category in orders:
        if category == "Electronics":
            customers_electronics_categories.add(cust_name)
        if category == "Clothing":
            customers_clothing_categories.add(cust_name)

customers_both_categories=customers_electronics_categories.intersection(customers_clothing_categories)

if(len(customers_both_categories)>=1):
    print("Customers who bought both Electronics and Clothing:")
    for cust_name in customers_both_categories:
        print(f"{cust_name}")

Customers who bought both Electronics and Clothing:
Akhilesh Kumar
Vipin Kumar


<h1><center>INSIGHTS</center></h1>

<h4>Extract the high-value customers and most frequently purchased products</h4>

From line 15 I've got below the top 3 customers. So we can give value add-on or more purchased to these customers by providing additional discounts or coupons

Adhir Samal → Total: 155
Ashish Pandey → Total: 110
Vipin Kumar → Total: 140


<h4>Identify trends based on category-wise sales </h4>

From line 10 I've got the below sales totals for each category wise. From this we can maintain inventory. Since Electronics is the top-performing category we can also use it as recommended products in cross sellings.

{'Electronics': 585, 'Clothing': 525, 'Home Essentials': 570}
